docs/architecture/commitlog/fileformat.rst

```diff
@@ -0,0 +1,60 @@
+ScyllaDB commitlog segment file format
+======================================
+
+File descriptor structure
+.........................
+
+ScyllaDB commitlog segment files are named with a versioned, time-indexed scheme, as::
+
+        <Prefix><version>-<id>.log
+
+Where `<Prefix>` is application specific, but typically "Commitlog-", or in case of 
+files being recycled "Recycled-Commitlog-", `<version>` is the file format version,
+and `<id>` is the id part of a replay position (timestamp + shard).
+
+Segment file data structure
+...........................
+
+All control data is written in network byte order.
+
+The file consists of a file header, followed by any number of chunks. Each chunk has 
+its own header + a marker to the start of next chunk, to allow skipping it more easily,
+should any data corruption be present in the chunk's data.
+
+Chunks contain data entries, with a small header, stored data + checksums to verify its integrity.
+
+An entry can be a "multi-entry", i.e. several entries written as one.
+
+Version 2
+---------
+
+(Format used in ScyllaDB 1.0 to as of this writing - named '2' because it is a slight deviation on the format used in cassandra)
+
+::
+
+        Segment file header
+
+        magic           : uint32_t - ('S'<<24) |('C'<< 16) | ('L' << 8) | 'C';
+        version         : uint32_t - same as descriptor
+        id              : uint64_t - same as descriptor
+        crc             : uint32_t - CRC32 of version, low 32 of id, high 32 of id. 
```

Comment:
Unrelated question. Presumably, if we lay out the bytes of int32_t we crc and then crc the entire buffer, we get the same result. But this depends on the byte order. Will crc work with network byte order or little-endian?

If network byte order, then to get then same "crc the objects or buffer, I don't care" behavior we should have first crc'ed the high 32 of id. If crc prefers little endian, then it won't work in the first place.

In the future, we should strive to allow crc of objects and buffers be the same.

Reply:
Our CRC32 does indeed process BE, so yes, indeed the ID crc is somewhat inconsistent. It is in fact inherited from cassandra IIRC. But I fail to see the significance of being fully consistent here (except for maybe allowing writing parser/verifier in one specific way). We are not treating this as a struct anyway (would be packed at the very least...)

Reply:
Just for consistency. If it's meticulously documented it doesn't add much.